lib/Frontend/CompileJobCacheKey.cpp

```
@@ -0,0 +1,81 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains utility methods for creating compile job cache keys.
+//
+//===----------------------------------------------------------------------===//
+
+#include <swift/Frontend/CompileJobCacheKey.h>
+#include <swift/Basic/Version.h>
+#include <llvm/ADT/SmallString.h>
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/CAS/HierarchicalTreeBuilder.h"
+#include "llvm/CAS/ObjectStore.h"
+
+using namespace swift;
+
+// TODO: Rewrite this into CASNodeSchema.
+llvm::Expected<llvm::cas::ObjectRef> swift::createCompileJobBaseCacheKey(
+    llvm::cas::ObjectStore &CAS, ArrayRef<const char *> Args) {
+  SmallString<256> CommandLine;
+
+  // TODO: Improve this list.
+  static const std::vector<std::string> removeArgAndNext = {
+      "-o", "-supplementary-output-file-map", "-serialize-diagnostics-path",
+      "-num-threads", "-cas-path"};
+
+  // Don't count the `-frontend` in the first location since only frontend
+  // invocation can have a cache key.
+  if (Args.size() > 1 && StringRef(Args.front()) == "-frontend")
+    Args = Args.drop_front();
+
+  bool SkipNext = false;
+  for (StringRef Arg : Args) {
+    if (SkipNext) {
+      SkipNext = false;
+      continue;
+    }
+    if (llvm::is_contained(removeArgAndNext, Arg)) {
+      SkipNext = true;
+      continue;
+    }
+    CommandLine.append(Arg);
+    CommandLine.push_back(0);
+  }
+
+  llvm::cas::HierarchicalTreeBuilder Builder;
+  Builder.push(
+      llvm::cantFail(CAS.storeFromString(None, CommandLine)),
```

This should report the Error. Same with the other cantFail below.